cmd/posix.go

```diff
@@ -1525,3 +1525,86 @@ func (s *posix) RenameFile(srcVolume, srcPath, dstVolume, dstPath string) (err e
 
 	return nil
 }
+
+func (s *posix) SpeedTestDrive(volume string) SpeedTestResultItem {
+	var err error
+	defer func() {
+		if err == errFaultyDisk {
+			atomic.AddInt32(&s.ioErrCount, 1)
+		}
+	}()
+
+	if atomic.LoadInt32(&s.ioErrCount) > maxAllowedIOError {
+		return SpeedTestResultItem{
+			Err: errFaultyDisk.Error(),
+		}
+	}
+
+	if err = s.checkDiskFound(); err != nil {
+		return SpeedTestResultItem{
+			Err: err.Error(),
+		}
+	}
+
+	// Verify if volume is valid and it exists.
+	volumeDir, err := s.getVolDir(volume)
+	if err != nil {
+		return SpeedTestResultItem{
+			Err: err.Error(),
+		}
+	}
+
+	// Validate effective path length before reading.
+	path := mustGetUUID()
+	filePath := pathJoin(volumeDir, path)
+	if err = checkPathLength((filePath)); err != nil {
+		return SpeedTestResultItem{
+			Err: err.Error(),
+		}
+	}
+
+	// Create a file with O_DIRECT flag
+	w, err := disk.OpenFileDirectIO(filePath, os.O_CREATE|os.O_WRONLY|os.O_EXCL, 0666)
+	if err != nil {
+		return SpeedTestResultItem{
+			Err: err.Error(),
+		}
+	}
+	// Remove the file created for speed test purposes
+	defer deleteFile(volumeDir, filePath)
+
+	// Fetch aligned buf for direct-io
+	buf := directio.AlignedBlock(speedTestBlockSize)
+
+	writeSpeed, err := speedTestWrite(w, buf)
+	w.Sync() // Sync before close
+	w.Close()
```

Comment:
@KP I got this while testing with the mc PR

![image](https://user-images.githubusercontent.com/5410427/60338559-ea4b8100-99c3-11e9-8b63-e7f3d602fa97.png)

I suspect that we are closing the response writer prematurely here